### ISSUE TITLE
\pgfmathparse incorrectly removes spaces and braces

### ISSUE THREAD

**Issue by domenjoud** [pgf-tikz/pgf#1077]
While parsing an expression `\pgfmathparse` incorrectly removes a space in second position and braces around a group in second position.

For instance, `e r` is read as `er`, `-{2,3}[1]` is read as `-2,3[1]`, and `a{bs}(1)`, `a{bs(1)}`, `a {b}s(1)`, `a {bs(1)}` or even `a {bs(}1)` are all read as `abs(1)`.  This happens only in second position: `pi r` is read correctly.

    \documentclass{article}
    
    \usepackage{pgf}
    
    \def\test#1{\pgfmathparse{#1}\typeout{#1 = \pgfmathresult}}
    
    \test{a bs(-1)}
    \test{a{bs}(-1)}
    \test{a {bs}(-1)}
    \test{a{}bs(-1)}
    \test{a {bs(}-1)}
    \test{e r}
    \test{-{2,3}[1]}
    
    \begin{document}
    \end{document}

results in

    a bs(-1) = 1.0
    a{bs}(-1) = 1.0
    a {bs}(-1) = 1.0
    a{}bs(-1) = 1.0
    a {bs(}-1) = 1.0
    
    ! Package PGF Math Error: Unknown function `er' (in 'e r').
    
    See the PGF Math package documentation for explanation.
    Type  H <return>  for immediate help.
     ...                                              
                                                      
    l.12 \test{e r}
                   
    ? 
    e r = 0
    
    ! LaTeX Error: Missing \begin{document}.
    
    See the LaTeX manual or LaTeX Companion for explanation.
    Type  H <return>  for immediate help.
     ...                                              
                                                      
    l.13 \test{-{2,3}[1]}

The problem has two distinct causes. The first one is the macro `\pgfmath@bgroup@strip` in the file 'pgfmathparser.code.tex' which removes the space in second position and the braces around a group in second position or third position after a space.

Rewriting this macro as follows

    \def\pgfmath@bgroup@strip#1{%
      \def\pgfmath@bgroup@stripped{#1}%
      \pgfmath@bgroup@@strip#1\pgfmath@
    }
    
    \def\pgfmath@bgroup@@strip{%
      \futurelet\pgfmath@token@let\pgfmath@bgroup@@@strip
    }
    
    \def\pgfmath@bgroup@@@strip{%
      \ifx\pgfmath@token@let\pgfmath@sptoken
        \expandafter\pgfmath@group@strip@sp
      \else
        \expandafter\pgfmath@group@@@@strip
      \fi
    }
    
    \expandafter\def\expandafter\pgfmath@group@strip@sp\pgfutil@space#1\pgfmath@{\pgfmath@bgroup@strip{#1}}
    
    \def\pgfmath@group@@@@strip#1\pgfmath@{%
      \def\pgfmath@bgroup@test{#1}%
      \ifx\pgfmath@bgroup@test\pgfmath@bgroup@stripped\else
        \pgfmath@bgroup@strip{#1}%
      \fi
    }

mostly solves the problem: `-{2,3}[1]` is parsed correctly and `a{bs}(1)` or  `a {bs(1)}` is reported as an error.

However, the correct expression `e r` is still read as `er`  because `\pgfmath@parse@function` (in pgfmathparser.code.tex) also removes the space in second position. This macro first discards any leading space. But when it is called from `\pgfmath@parse@@operand@` the first character of the function name has already been read by `\pgfmath@parse@@operand@` and put in `\pgfmath@function`. Therefore, ignoring leading spaces is incorrect.

A solution might be to modify `\pgfmath@parse@@operand@` and reinject this first character in the input instead of putting it in `\pgfmath@function`:

    % An operand is
    % * |+| (unary plus),
    % * |-| (unary minus),
    % * a prefix operator
    \def\pgfmath@parse@@operand@#1{%
      \def\pgfmath@token{#1}%
      \ifx\pgfmath@token\pgfmath@char@plus
        % Unary (prefix) plus
        \let\pgfmath@parse@next=\pgfmath@parse@@operand%
      \else
        \ifx\pgfmath@token\pgfmath@char@minus
          % Unary (prefix) minus
          \pgfmath@stack@push@operation{neg}%
          \let\pgfmath@parse@next=\pgfmath@parse@@operand
        \else
          % See |\pgfmathdeclareoperator| (below). Note that |+| and |-|
          % that are also declared as operators have been tested before so
          % they should never be parsed as binary (infix) operator at this stage.
          \expandafter\ifx\csname pgfmath@operation@\expandafter\string\pgfmath@token @prefix\endcsname\pgfmath@token
            \let\pgfmath@parse@next=\pgfmath@parse@prefix@operator
          \else
            \ifpgfmath@quickparse
              \let\pgfmath@parse@next=\pgfmath@qparse@operand
            \else
              \let\pgfmath@number=\pgfmath@empty
              \expandafter\ifx\csname pgfmath@token@functional@\string#1\endcsname\relax%
                \let\pgfmath@base=\pgfmath@empty%
                \let\pgfmath@token@next=\pgfmath@token%
                \if#10% Check for octal prefix.
                  \ifpgfmath@octalparsing%
                    \def\pgfmath@base{8}%
                  \fi%
                \fi%
                \let\pgfmath@parse@next=\pgfmath@parse@number%
              \else%
                %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
                %% replace the following two lines  
                %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
                % \let\pgfmath@function=\pgfmath@token%
                % \let\pgfmath@parse@next=\pgfmath@parse@function%
                %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
                %% with these two
                %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
                \let\pgfmath@function=\pgfmath@empty
                \def\pgfmath@parse@next{\pgfmath@parse@function#1}%
              \fi%
            \fi%
          \fi%
        \fi%
      \fi%
      \pgfmath@parse@next%
    }

**Comment by muzimuzhi** [comment]
> ```tex
> \expandafter\def\expandafter\pgfmath@group@strip@sp\pgfutil@space#1\pgfmath@{\pgfmath@bgroup@strip{#1}}
> ```

This will make the parser accept input like `\pgfmathparse{{ { e}}}`.

We may borrow some ideas of [`expkv`'s `\ekv@strip`][ekv@strip] if it's safe to have `\pgfmathparse` strip exactly one level of braces after stripping at most one space from each side of `<math expr>`.

[ekv@strip]: https://github.com/Skillmon/tex_expkv/blob/5106175fcbda1a05aa84b6b22459f4dc74d2ad54/expkv.dtx#L1815-L1839

**Comment by domenjoud** [comment]
Yes, this input would be accepted. But I fail to find a situation where it would cause a problem. Since ' ...' and '{...}' are both considered equivalent to '...' it seems coherent to apply it recursively and consider '{{ { e}}}' equivalent to 'e'.

**Comment by ilayn** [comment]
What would be the use case for `\pgfmathparse{e r}` while you have `\pgfmathparse{rad(e)}` ? Note that `r` suffix is a TikZ convenience for quick angle referencing. It is not really meant to enhance units handling of pgfmath lib. I don't see why any of these should be accepted as valid input. It's nice that they work but in a rather C's undefined behavior type of way.

**Comment by domenjoud** [comment]
' e r' is a correct syntax described in the pgf manuel and therefore it should work as expected. And as I mentioned, the problem is worse than just the case of 'e r'.  By the way, 'e r' is deg(e)' and not 'rad(e)'.

**Comment by ilayn** [comment]
> And as I mentioned, the problem is worse than just the case of 'e r'.

Yes but what is that worse case? You have only provided pathological cases where I would consider them as typos.

>  By the way, 'e r' is deg(e)' and not 'rad(e)'.

Ah, sure.

**Comment by domenjoud** [comment]
I don't really see your point. There is clearly a bug and I proposed a solution. An erroneous input should be reported as such and a correct input should be handled as expected.

**Comment by ilayn** [comment]
I am seeing here is `e r` the only bug. It is much easier to use `deg(e)` anyways and like I mentioned it is just a shortcut. Removing `e r` from the manual is much easier and putting it to higher up for drawing purposes is much easier.

And all the rest is just brace hunting for no apparent use case. That's my point.

**Comment by muzimuzhi** [comment]
If we take `pgfmath` as a real/serious math engine and expect it to correctly evaluate all valid math expressions and throw errors on invalid ones, then `e r`, or always dropping the second token if it's a space, _is_ a bug. To give more (wired and avoidable) examples, consider
```tex
\pgfmathdeclarefunction{x}{1}{\def\pgfmathresult{#1}}
% valid math expr, should return `x(10) = 10`, but throws error
\pgfmathparse{x 10}\pgfmathresult

\pgfmathdeclarefunction{xx}{1}{\def\pgfmathresult{#1}}
% invalid math expr, should throw an error "unknown function `x'", but evaluates it as `xx 10 = xx(10) = 10`
\pgfmathparse{x x 10}\pgfmathresult
```

**Comment by ilayn** [comment]
Well to start with it is not nearly a serious math engine. We are talking about TeX here so let's not get carried away. Secondly, `e r` is as with any considerable math engine is just horrific syntax. Don't confuse a one-off convenience with a proper math parsing. Couple this with TeX in-place replacement of variables and you have all kinds of disasters ready for you. 

When you encounter with such nonsensical results coming out of a system, the solution is always going with the simpler rule and not code 67 different edge cases.

**Comment by ilayn** [comment]
Those wired and avoidable cases are exactly what I mean by undefined behavior candidates. If you write such math there is no system in the world that would give you sensible results other than CAS systems where they implement guessing rules. 

That's why I am asking for a use case. And the use case provided here is to skip keystrokes so far and not a valid statement giving wrong answers.

**Comment by domenjoud** [comment]
@muzimuzhi I did not notice that parentheses are optional for unary fonctions. For my part, I would not expect `f 10` to be parsed as `f(10)`. I played with this and notice that `gcd 10,15` evaluates as `{10}{15}` without reporting an error.

**Comment by muzimuzhi** [comment]
@domenjoud I also didn't notice that feature until your issue pushes me think of more edge/test cases and then read the manual and find (line 409):
https://github.com/pgf-tikz/pgf/blob/d835e0312028e7e398626bdd77e5effeaab280f0/doc/generic/pgf/text-en/pgfmanual-en-math-parsing.tex#L403-L412

**Comment by domenjoud** [comment]
Usually a good idea to read the manual, I should do this more often :-). But this does not explain why `gcd 46` yields `2`, which means that it is parsed as `gcd(4,6)`. My previous example was wrong: `gcd 10,15` yields `{1}{15}` which means that it is parsed as `gcd(1,0),15`. Strange and unexpected!

**Comment by AlainMatthes** [comment]
> I am seeing here is `e r` the only bug. It is much easier to use `deg(e)` anyways and like I mentioned it is just a shortcut. Removing `e r` from the manual is much easier and putting it to higher up for drawing purposes is much easier.
> 
> And all the rest is just brace hunting for no apparent use case. That's my point.

If  ` \pgfmathparse{pi r}\pgfmathresult ` is correct then

 `\def\a{pi} \pgfmathparse{\a r}\pgfmathresult`
gives an error.

And there is a real problem with   `  \def\x{pi}  \tikz \draw (1,0) -- ({cos(\x r)}, {sin(\x r)})  ;`

I used some tricks to get around this difficulty like

`def\x{pi\space} `or ` \def\x{pi/1}` 

If `{e r}` n'est pas très courant `pi r` is more annoying

**Comment by muzimuzhi** [comment]
@AlainMatthes Since your problem is mostly not relevant to the current issue, I've added an answer directly under your question on TeX-SX: https://tex.stackexchange.com/a/641185. Sorry for the misleading info I provided there in question comments.

**Comment by domenjoud** [comment]
By the way, as I already mentioned, the problem with this bug goes further than the simple case of `e r`. Even if `r` is removed from pgf as suggested, I find still annoying that an incorrect expression as `c os(x)` is parsed as `cos(x)`. And worse, the brace is stripped in the correct expression `-{1,2}[1]` which is then parsed as `-1,2[1]` and causes an error.
